File: sdks/py/src/jeffs_brain_memory/llm/fake.py
# SPDX-License-Identifier: Apache-2.0
"""Deterministic in-memory provider and embedder for tests and dev."""

from __future__ import annotations

import hashlib
import math
import struct
from typing import AsyncIterator

from .types import (
    CompleteRequest,
    CompleteResponse,
    EmptyMessagesError,
    StopReason,
    StreamChunk,
)


class FakeProvider:
    """Hand out responses from a fixed list in round-robin order."""

    def __init__(self, responses: list[str]) -> None:
        self._responses: list[str] = list(responses)
        self._idx = 0
        self._closed = False

    def _next(self) -> str:
        if not self._responses:
            return ""
        value = self._responses[self._idx % len(self._responses)]
        self._idx += 1
        return value

    async def complete(self, req: CompleteRequest) -> CompleteResponse:
        if self._closed:
            raise RuntimeError("llm: fake provider closed")
        if not req.messages:
            raise EmptyMessagesError()
        text = self._next()
        tokens_in = sum(len(m.content) for m in req.messages)
        return CompleteResponse(
            text=text,
            stop=StopReason.END_TURN,
            tokens_in=tokens_in,
            tokens_out=len(text),
        )

    async def complete_stream(self, req: CompleteRequest) -> AsyncIterator[StreamChunk]:
        if self._closed:
            raise RuntimeError("llm: fake provider closed")
        if not req.messages:
            raise EmptyMessagesError()
        text = self._next()
        return _fake_stream(text)

    async def close(self) -> None:
        self._closed = True


async def _fake_stream(text: str) -> AsyncIterator[StreamChunk]:
    for ch in text:
        yield StreamChunk(delta_text=ch)
    yield StreamChunk(stop=StopReason.END_TURN)


class FakeEmbedder:
    """Emit deterministic unit-length vectors seeded by SHA-256 of each input."""

    def __init__(self, dims: int) -> None:
        self._dims = dims if dims > 0 else 16

    async def embed(self, texts: list[str]) -> list[list[float]]:
        return [_seed_vector(t, self._dims) for t in texts]

    def dimensions(self) -> int:
        return self._dims

    async def close(self) -> None:
        return None


def _seed_vector(text: str, dims: int) -> list[float]:
    if dims == 0:
        return []
    seed = hashlib.sha256(text.encode("utf-8")).digest()
    vec: list[float] = []
    sum_sq = 0.0
    for i in range(dims):
        offset = (i * 4) % len(seed)
        end = offset + 4
        if end > len(seed):
            seed = hashlib.sha256(seed + bytes([i & 0xFF])).digest()
            offset = 0
            end = 4
        raw = struct.unpack(">I", seed[offset:end])[0]
        val = (raw / 0xFFFFFFFF) * 2.0 - 1.0
        vec.append(val)
        sum_sq += val * val
    if sum_sq == 0.0:
        return vec
    norm = math.sqrt(sum_sq)
    return [v / norm for v in vec]

File: go/llm/fake.go
// SPDX-License-Identifier: Apache-2.0

package llm

import (
	"context"
	"crypto/sha256"
	"encoding/binary"
	"math"
	"strings"
	"sync"
	"sync/atomic"
)

// NewFake returns a [Provider] that hands out responses from the provided
// slice in round-robin order. It is deterministic and requires no network.
// This is the provider downstream packages use in unit tests.
func NewFake(responses []string) Provider {
	copyOf := make([]string, len(responses))
	copy(copyOf, responses)
	return &fakeProvider{responses: copyOf}
}

type fakeProvider struct {
	responses []string
	idx       atomic.Uint64
	closed    atomic.Bool
}

func (f *fakeProvider) next() string {
	if len(f.responses) == 0 {
		return ""
	}
	i := f.idx.Add(1) - 1
	return f.responses[int(i%uint64(len(f.responses)))]
}

func (f *fakeProvider) Complete(ctx context.Context, req CompleteRequest) (CompleteResponse, error) {
	if f.closed.Load() {
		return CompleteResponse{}, context.Canceled
	}
	if err := ctx.Err(); err != nil {
		return CompleteResponse{}, err
	}
	if len(req.Messages) == 0 {
		return CompleteResponse{}, ErrEmpty
	}
	text := f.next()
	return CompleteResponse{
		Text:      text,
		Stop:      StopEndTurn,
		TokensIn:  len(concatUserText(req.Messages)),
		TokensOut: len(text),
	}, nil
}

func (f *fakeProvider) CompleteStream(ctx context.Context, req CompleteRequest) (<-chan StreamChunk, error) {
	if f.closed.Load() {
		return nil, context.Canceled
	}
	if len(req.Messages) == 0 {
		return nil, ErrEmpty
	}
	text := f.next()
	ch := make(chan StreamChunk, len(text)+1)
	go func() {
		defer close(ch)
		for _, r := range text {
			select {
			case <-ctx.Done():
				return
			case ch <- StreamChunk{DeltaText: string(r)}:
			}
		}
		select {
		case <-ctx.Done():
		case ch <- StreamChunk{Stop: StopEndTurn}:
		}
	}()
	return ch, nil
}

func (f *fakeProvider) Close() error {
	f.closed.Store(true)
	return nil
}

func concatUserText(msgs []Message) string {
	var b strings.Builder
	for _, m := range msgs {
		b.WriteString(m.Content)
	}
	return b.String()
}

// NewFakeEmbedder returns an [Embedder] that emits deterministic pseudo-
// random vectors seeded by a SHA-256 hash of each input string. This lets
// tests verify embedding-driven logic without touching a real model.
func NewFakeEmbedder(dims int) Embedder {
	if dims <= 0 {
		dims = 16
	}
	return &fakeEmbedder{dims: dims}
}

type fakeEmbedder struct {
	dims int
	mu   sync.Mutex
}

func (f *fakeEmbedder) Embed(ctx context.Context, texts []string) ([][]float32, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	out := make([][]float32, len(texts))
	for i, t := range texts {
		out[i] = seedVector(t, f.dims)
	}
	return out, nil
}

func (f *fakeEmbedder) Dimensions() int { return f.dims }

func (f *fakeEmbedder) Close() error { return nil }

// seedVector produces a unit-length deterministic vector for t.
func seedVector(t string, dims int) []float32 {
	vec := make([]float32, dims)
	if dims == 0 {
		return vec
	}
	seed := sha256.Sum256([]byte(t))
	var sumSq float64
	for i := 0; i < dims; i++ {
		offset := (i * 4) % len(seed)
		end := offset + 4
		if end > len(seed) {
			// Re-hash for bigger dims.
			nextSeed := sha256.Sum256(append(seed[:], byte(i)))
			seed = nextSeed
			offset = 0
			end = 4
		}
		raw := binary.BigEndian.Uint32(seed[offset:end])
		val := float64(raw)/float64(math.MaxUint32)*2 - 1
		vec[i] = float32(val)
		sumSq += val * val
	}
	if sumSq == 0 {
		return vec
	}
	norm := float32(math.Sqrt(sumSq))
	for i := range vec {
		vec[i] /= norm
	}
	return vec
}
